File: scripts/puzzle/binary/generate.go
package binary

import (
	"errors"
	"math/rand"
)

type lineState struct {
	counts cellCounts
	bits   uint32 // Limits size to 32 * 32
}

func (s *lineState) add(value uint8, bitIdx int) {
	s.counts.add(value)
	if value == 1 {
		s.bits |= uint32(1) << bitIdx
	}
}

func (s *lineState) remove(value uint8, bitIdx int) {
	switch value {
	case 0:
		s.counts.zeros--
	case 1:
		s.counts.ones--
		s.bits &^= uint32(1) << bitIdx
	}
}

// generate Generates a new puzzle
func (b *Binary) generate() error {
	// Reset the field
	b.Field = make([]uint8, b.Size*b.Size)

	// Get all possible rows
	rows := generateRows(b.Size)
	if len(rows) == 0 {
		return errors.New("no possible rows")
	}

	// Shuffle rows
	rand.Shuffle(len(rows), func(i, j int) {
		rows[i], rows[j] = rows[j], rows[i]
	})

	usedRows := make([]bool, len(rows))
	states := make([]lineState, b.Size)
	g := newGrid(b.Size, b.Field)

	// Try to fill in field with rows
	ok := b.fillRows(g, rows, usedRows, states, 0)
	if !ok {
		return errors.New("unable to fill field")
	}

	return nil
}

func (b *Binary) fillRows(g grid, candidates [][]uint8, usedRows []bool, states []lineState, rowIdx int) bool {
	if rowIdx == g.size {
		return true
	}

	// Keep trying every unused candidate until one leads to a solution
	for idx, candidate := range candidates {
		// Already used
		if usedRows[idx] {
			continue
		}

		if !canPlaceRow(g, states, rowIdx, candidate) {
			continue
		}

		g.setRow(rowIdx, candidate)
		applyRow(states, candidate, rowIdx) // Keeps track of columns
		usedRows[idx] = true

		ok := true
		// If it is the last row check for column uniqueness
		// Row uniques is guaranteed by how the rows are generated
		if rowIdx == g.size-1 && !columnsUnique(states) {
			ok = false
		}

		// Go deeper
		if ok && b.fillRows(g, candidates, usedRows, states, rowIdx+1) {
			return true
		}

		// Unable to complete the field
		// Try another row
		usedRows[idx] = false
		undoRow(states, candidate, rowIdx)
	}

	return false
}

// canPlaceRow checks if placing the row won't cause any immediate conflicts
func canPlaceRow(g grid, states []lineState, rowIdx int, candidate []uint8) bool {
	size := g.size
	half := size / 2
	remainingRows := size - rowIdx - 1

	for col, state := range states {
		bit := candidate[col]
		counts := state.counts
		counts.add(bit)

		if !counts.canBalanceWithRemaining(half, remainingRows) {
			return false
		}

		// No triplets
		if rowIdx >= 2 {
			a := g.at(rowIdx-2, col)
			prev := g.at(rowIdx-1, col)
			if bit == a && bit == prev {
				return false
			}
		}
	}

	return true
}

// applyRow updates the affected columns state if a row is added
func applyRow(states []lineState, candidate []uint8, rowIdx int) {
	for col, bit := range candidate {
		states[col].add(bit, rowIdx)
	}
}

// undoRow updates the affected columns state if a row is removed
func undoRow(states []lineState, candidate []uint8, rowIdx int) {
	for col, bit := range candidate {
		states[col].remove(bit, rowIdx)
	}
}

// columnsUnique checks if every column is unique
func columnsUnique(states []lineState) bool {
	// The bit state to determine if a column is equal is maintained by applyRow and undoRow
	seen := make(map[uint32]bool, len(states))
	for _, s := range states {
		if _, ok := seen[s.bits]; ok {
			return false
		}
		seen[s.bits] = true
	}

	return true
}

// generateRows generates all possible rows for a given size
// It only generates valid and unique rows
func generateRows(size int) [][]uint8 {
	half := size / 2
	out := make([][]uint8, 0)

	buf := make([]uint8, size)

	var gen func(col, zeros, ones int)
	gen = func(col, zeros, ones int) {
		// Max zeros and ones == half of the size
		if zeros > half || ones > half {
			return
		}

		// Is reaching half even still possible?
		remaining := size - col
		if zeros+remaining < half || ones+remaining < half {
			return
		}

		// No triplets
		if col >= 3 {
			if buf[col-3] == buf[col-2] && buf[col-2] == buf[col-1] {
				return
			}
		}

		// Full row
		if col == size {
			if zeros == half && ones == half {
				cpy := make([]uint8, size)
				copy(cpy, buf)
				out = append(out, cpy)
			}
			return
		}

		buf[col] = 0
		gen(col+1, zeros+1, ones)

		buf[col] = 1
		gen(col+1, zeros, ones+1)
	}

	gen(0, 0, 0)
	return out
}

File: backend/src/puzzle/takuzu/generate.rs
use crate::puzzle::takuzu::TakuzuCell;
use crate::puzzle::takuzu::TakuzuCell::{X, O, Empty};


#[derive(Clone, Copy, Default)]
struct CellCounts {
    x_count: u8,
    o_count: u8,
}

impl CellCounts {
    fn add(&mut self, value: TakuzuCell) {
        match value {
            X => self.x_count += 1,
            O => self.o_count += 1,
            _ => {}
        }
    }

    fn remove(&mut self, value: TakuzuCell) {
        match value {
            X => self.x_count -= 1,
            O => self.o_count -= 1,
            _ => {}
        }
    }

    fn can_balance_with_remaining(&self, half: u8) -> bool {
        self.x_count <= half
            && self.o_count <= half
    }
}

#[derive(Clone, Copy, Default)]
struct LineState {
    counts: CellCounts,
    bits: u32,
}


impl LineState {
    fn add(&mut self, value: TakuzuCell, bit_idx: usize) {
        self.counts.add(value);
        if value == X {
            self.bits |= 1 << bit_idx;
        }
    }

    fn remove(&mut self, value: TakuzuCell, bit_idx: usize) {
        self.counts.remove(value);
        if value == X {
            self.bits &= !(1 << bit_idx);
        }
    }
}

pub(super) fn generate_grid(size: u8) -> Vec<Vec<TakuzuCell>> {
    let mut grid = vec![vec![O; size as usize]; size as usize];

    let mut rows = generate_possible_rows(size);

    // Optional shuffle
    use rand::seq::SliceRandom;
    let mut rng = rand::thread_rng();
    rows.shuffle(&mut rng);

    let mut used_rows = vec![false; rows.len()];
    let mut states = vec![LineState::default(); size as usize];

    fill_rows(
        &mut grid,
        &rows,
        &mut used_rows,
        &mut states,
        0,
        size,
    );

    grid
}

fn fill_rows(
    grid: &mut Vec<Vec<TakuzuCell>>,
    candidates: &[Vec<TakuzuCell>],
    used_rows: &mut [bool],
    states: &mut [LineState],
    row_idx: usize,
    size: u8,
) -> bool {
    if row_idx == size as usize {
        return true;
    }

    for (idx, candidate) in candidates.iter().enumerate() {
        if used_rows[idx] {
            continue;
        }

        if !can_place_row(grid, states, row_idx, candidate, size) {
            continue;
        }

        grid[row_idx] = candidate.clone();
        apply_row(states, candidate, row_idx);
        used_rows[idx] = true;

        let mut ok = true;

        if row_idx == size as usize - 1 && !columns_unique(states) {
            ok = false;
        }

        if ok && fill_rows(grid, candidates, used_rows, states, row_idx + 1, size) {
            return true;
        }

        used_rows[idx] = false;
        undo_row(states, candidate, row_idx);
    }

    false
}

fn can_place_row(
    grid: &[Vec<TakuzuCell>],
    states: &[LineState],
    row_idx: usize,
    candidate: &[TakuzuCell],
    size: u8,
) -> bool {
    let half = size / 2;

    for (col, state) in states.iter().enumerate() {
        let bit = candidate[col];

        let mut counts = state.counts;
        counts.add(bit);

        if !counts.can_balance_with_remaining(half) {
            return false;
        }

        // No vertical triplets
        if row_idx >= 2 {
            let a = grid[row_idx - 2][col];
            let b = grid[row_idx - 1][col];
            if bit == a && bit == b {
                return false;
            }
        }
    }

    true
}

fn apply_row(states: &mut [LineState], candidate: &[TakuzuCell], row_idx: usize) {
    for (col, &cell) in candidate.iter().enumerate() {
        states[col].add(cell, row_idx);
    }
}

fn undo_row(states: &mut [LineState], candidate: &[TakuzuCell], row_idx: usize) {
    for (col, &cell) in candidate.iter().enumerate() {
        states[col].remove(cell, row_idx);
    }
}

fn columns_unique(states: &[LineState]) -> bool {
    use std::collections::HashSet;

    let mut seen = HashSet::with_capacity(states.len());

    for state in states {
        if !seen.insert(state.bits) {
            return false;
        }
    }

    true
}
fn generate_possible_rows(size: u8) -> Vec<Vec<TakuzuCell>> {
    let mut possible_rows: Vec<Vec<TakuzuCell>> = vec![];


    let mut current_row: Vec<TakuzuCell> = vec![Empty; size as usize];
    fn generate_row (
        row_index: usize,
        current_x_count: u8,
        current_o_count: u8,
        current_row: &mut Vec<TakuzuCell>,
        possible_rows: &mut Vec<Vec<TakuzuCell>>,
        size: u8
    ) {
        let half_size = size / 2;
        if current_o_count > half_size || current_x_count > half_size {
            return;
        }

        if row_index >= 3 {
            if current_row[row_index - 3] == current_row[row_index - 2]
                && current_row[row_index - 2] == current_row[row_index - 1] {
                return;
            }
        }

        if row_index == usize::from(size) {
            if current_x_count == half_size && current_o_count == half_size {
                possible_rows.push(current_row.to_vec());
            }
            return;
        }

        current_row[row_index] = X;
        generate_row(row_index + 1, current_x_count + 1, current_o_count, current_row, possible_rows, size);

        current_row[row_index] = O;
        generate_row(row_index + 1, current_x_count, current_o_count + 1, current_row, possible_rows, size);
    }
    generate_row(0, 0, 0, &mut current_row, &mut possible_rows, size);
    possible_rows
}